crates/models/codegen/src/lib.rs

```diff
@@ -0,0 +1,427 @@
+            params,
+            self.hyperparameters.n_ctx,
+            self.hyperparameters.n_layer,
+            self.hyperparameters.n_embd,
+            self.hyperparameters.n_vocab,
+        )
+    }
+
+    fn evaluate(
+        &self,
+        session: &mut InferenceSession,
+        params: &InferenceParameters,
+        input_tokens: &[TokenId],
+        output_request: &mut EvaluateOutputRequest,
+    ) {
+        let n = input_tokens.len();
+        let n_threads = params.n_threads;
+
+        let Hyperparameters {
+            n_embd,
+            n_head,
+            n_vocab,
+            n_layer,
+            n_rot,
+            ..
+        } = self.hyperparameters;
+        let n_ctx = self.n_context_tokens;
+
+        let (ctx0, embd) = common::prepare_for_evaluate(n_layer, session, input_tokens);
+
+        let n_past = session.n_past;
+
+        // wte
+        let mut input_layer = ctx0.op_get_rows(&self.wte, &embd);
+
+        let memory_k = &session.memory_k;
+        let memory_k_size = memory_k.element_size();
+
+        let memory_v = &session.memory_v;
+        let memory_v_size = memory_v.element_size();
+
+        let mut gf = ggml::ComputationGraph::new(n_threads);
+
+        for il in 0..n_layer {
+            // norm
+            let mut current = ctx0.op_norm(&input_layer);
+            current = ctx0.op_add(
+                &ctx0.op_mul(&ctx0.op_repeat(&self.layers[il].ln_1_g, &current), &current),
+                &ctx0.op_repeat(&self.layers[il].ln_1_b, &current),
+            );
+
+            let input_sa = current.share();
+
+            // the number of logical TPU-v4 cores, is 4 so mp_num must be a factor of 4
+            let mp_num = 4;
+
+            // TODO!: rewrite self attention for codegen qkv vector
+            // TODO!: break up qkv vector into individual q, k ,v
+            let qkv_cur = ctx0.op_reshape_3d(&self.layers[il].c_attn_qkv_proj_w, n_embd / n_head, n_head, n_past + n);
+
+            let local_dim = n_embd / mp_num;
+
+            let (q, k, v) =
+
+
+            //let qcur = ctx0.op_rope(
+                //&ctx0.op_reshape_3d(
+                    //&ctx0.op_mul_mat(&self.layers[il].c_attn_q_proj_w, &current),
+                    //n_embd / n_head,
+                    //n_head,
+                    //n,
+                //),
+                //n_past,
+                //n_rot,
+                //0,
+            //);
+            //let kcur = ctx0.op_rope(
+                //&ctx0.op_reshape_3d(
+                    //&ctx0.op_mul_mat(&self.layers[il].c_attn_k_proj_w, &current),
+                    //n_embd / n_head,
+                    //n_head,
+                    //n,
+                //),
+                //n_past,
+                //n_rot,
+                //0,
+            //);
+//
+            //// self-attention store key and value to memory
+            //let vcur =
+                //ctx0.op_transpose(&ctx0.op_mul_mat(&self.layers[il].c_attn_v_proj_w, &current));
+//
```

Comment:
I don't really love checking this kind of stuff into a `main` branch. Can this be removed? If something needs to be persisted for posterity, maybe it's more appropriate to use an Issue?

Reply:
Yeah I will remove that. I'm just in the middle of finding out how to best implement the qkv vector implementations.